lib/client/db/mcp/managedconn.go

```diff
@@ -0,0 +1,164 @@
+package mcp
+
+import (
+	"context"
+	"log/slog"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+)
+
+// ManagedConn manages a database connection, providing a way to keep a single
+// active connection per database, and also closing the connection once it
+// becomes inactive.
+//
+// Managed connections are always used with Exec function.
+type ManagedConn[T any, C conn[T]] struct {
+	mu   sync.Mutex
+	cond sync.Cond
+
+	cfg      *ManagedConnConfig
+	newFunc  func(context.Context) (*T, error)
+	active   C
+	closed   bool
+	watchdog *time.Timer
+
+	// cancelExec is the context cancelation function of a execution. It will
+	// only be set when there is a running execution, so it can be used to
+	// indicate the connection is in use.
+	cancelExec context.CancelFunc
```

Comment:
I didn't understand this comment without checking out the rest of code. Perhaps a slight change?

```suggestion
	// cancelExec will cancel currently executed query when invoked. It will
	// only be set when there is a running execution, so it can be also used to
	// indicate the connection is in use.
	cancelExec context.CancelFunc
```

Also, looking at usages around the code, I kind of wonder if having a thread-safe context with lifetime tied to currently active query wouldn't work better. Just an vague idea here, feel free to disregard.